rotkehlchen/tests/data_migrations/test_migration_20.py

```diff
@@ -345,3 +344,228 @@ def test_migration_20_edge_case_multiple_swaps(database: DBHandler) -> None:
+    backup_db_path = database.user_data_dir / f'{ts_now()}_rotkehlchen_db_v47.backup'
+    backup_conn = DBConnection(
+        path=str(backup_db_path),
+        connection_type=DBConnectionType.USER,
+        sql_vm_instructions_cb=database.sql_vm_instructions_cb,
+    )
+    unlock_database(
+        db_connection=backup_conn,
+        password=database.password,
+        sqlcipher_version=database.sqlcipher_version,
+    )
+    script = f"PRAGMA key='{protect_password_sqlcipher(database.password)}';"
+    if database.sqlcipher_version == 3:
+        script += f'PRAGMA kdf_iter={KDF_ITER};'
+
+    backup_conn.executescript(script)
+    backup_conn.execute('PRAGMA foreign_keys=ON')
+    with backup_conn.write_ctx() as cursor:  # recreate the trades table schema from v47.
+        cursor.execute("""
+            CREATE TABLE IF NOT EXISTS trades (
+                id INTEGER PRIMARY KEY,
+                timestamp INTEGER NOT NULL,
+                location TEXT NOT NULL,
+                base_asset TEXT NOT NULL,
+                quote_asset TEXT NOT NULL,
+                type TEXT NOT NULL,
+                amount TEXT NOT NULL,
+                rate TEXT NOT NULL,
+                fee TEXT,
+                fee_currency TEXT,
+                link TEXT,
+                notes TEXT
+            )
+        """)
+
+        # insert trades that would be affected by the v1.39.0 bug.
+        # these trades have empty link fields and share locations, which caused ID conflicts.
+        cursor.executemany(
+            """INSERT INTO trades (timestamp, location, base_asset, quote_asset, type,
+           amount, rate, fee, fee_currency, link, notes)
+           VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
+            """,
+            [
+                # Two trades at same timestamp and location - these would conflict
+                (1000000, 'E', 'BTC', 'USD', 'A', '0.5', '40000', '0.001', 'BTC', '', 'First trade'),  # noqa: E501
+                (1000000, 'E', 'ETH', 'USD', 'A', '1.5', '2000', '0.01', 'ETH', '', 'Second trade'),  # noqa: E501
+                # Trade with no fee information
+                (2000000, 'B', 'ETH', 'USD', 'B', '2.0', '1800', None, None, '', 'Third trade'),
+                # Settlement trade type
+                (3000000, 'G', 'BTC', 'EUR', 'C', '0.25', '35000', '5', 'EUR', '', 'Settlement'),
+            ],
+        )
+
+    backup_conn.close()
+
+    db_events = DBHistoryEvents(database)  # add event with location hash identifier to trigger recovery  # noqa: E501
+    trigger_event = _create_location_hash_event(Location.BINANCE, (binance_timestamp_ms := TimestampMS(1000000000)))  # noqa: E501
+    with database.user_write() as write_cursor:
+        db_events.add_history_events(write_cursor, history=[trigger_event])
+
+    with patch('rotkehlchen.data_migrations.manager.MIGRATION_LIST', new=[MIGRATION_LIST[10]]):  # Migration 20 is at index 10  # noqa: E501
```

Comment:
So the test should start from here.  The data should just be in the DB. But it should verify what exists in the DB. So the trades. This is just to save space in our test code